tools/buildsys/src/builder.rs

```diff
@@ -255,6 +281,106 @@ where
         .collect()
 }
 
+/// Create a directory for build artifacts.
+fn create_build_dir(kind: &BuildType, name: &str) -> Result<PathBuf> {
+    let prefix = match kind {
+        BuildType::Package => "packages",
+        BuildType::Variant => "variants",
+    };
+
+    let path = [&getenv("BUILDSYS_STATE_DIR")?, prefix, name]
+        .iter()
+        .collect();
+
+    fs::create_dir_all(&path).context(error::DirectoryCreate { path: &path })?;
+
+    Ok(path)
+}
+
+/// Copy build artifacts to the output directory.
+fn copy_build_files<P>(build_dir: P, output_dir: P) -> Result<()>
+where
+    P: AsRef<Path>,
+{
+    fn is_artifact(entry: &DirEntry) -> bool {
+        entry.file_type().is_file()
+            && entry
+                .file_name()
+                .to_str()
+                .map(|s| !s.ends_with(".marker"))
+                .unwrap_or(false)
+    }
+
+    for artifact_file in find_files(&build_dir, is_artifact) {
+        let mut marker_file = artifact_file.clone().into_os_string();
+        marker_file.push(".marker");
+        fs::File::create(&marker_file).context(error::FileCreate { path: &marker_file })?;
+
+        let mut output_file: PathBuf = output_dir.as_ref().into();
+        output_file.push(
+            artifact_file
+                .file_name()
+                .context(error::BadFilename { path: &output_file })?,
+        );
+
+        fs::rename(&artifact_file, &output_file)
+            .context(error::FileRename { path: &output_file })?;
+    }
+
+    Ok(())
+}
+
+/// Remove build artifacts from the output directory.
+fn clean_build_files<P>(build_dir: P, output_dir: P) -> Result<()>
+where
+    P: AsRef<Path>,
+{
+    fn is_marker(entry: &DirEntry) -> bool {
+        entry
+            .file_name()
+            .to_str()
+            .map(|s| s.ends_with(".marker"))
```

Comment:
```suggestion
            .map(|s| s.ends_with(".buildsys_marker"))
```
In case a package has a real `bla.marker` file...

On that note, it took me a while to understand what we're tracking here and the assumptions we're making.  I think we can only handle one level of files, no subdirectories, because we're not tracking anything about the path within the build / output directories.  I think it'd be worth documenting more of what's going on, example use cases, etc.